Add "reset password via SMS" workflow

jonboiser opened the pull request:
**Providing Phone Number to Get SMS**
![get token](https://user-images.githubusercontent.com/10248067/31907495-d94a168c-b7e8-11e7-8e87-91f93ba1856c.gif)

**Resetting password**
![reset pw](https://user-images.githubusercontent.com/10248067/31907502-ddb2e15e-b7e8-11e7-9ff3-5d0f92b11e15.gif)


To test:

1. Install the SMS simulator (see the README in the `/smpp` folder of this repo)
1. Go to http://localhost:8000/user/api/phonesignup/ and create an account with your phone number as `username`. IMPORTANT! Save the actual username (e.g. `9ff4e87a0673406191ffe8b0a37777` that will appear as soon as the account is created.)
1. Go to http://localhost:8000/user/#/signin
1. Click "Reset password"
1. Happy-path: (Hopefully) go through the dialogue without any trouble
1. Non-happy paths: a) enter a bogus phone number, b) enter a legit phone number, but with SMPP simulator turned off. Should get some explanation from UI why reset attempt didn't work.
1. If finished happy path, go to terminal window with SMPP and look for a message like this: `please click the following link: http://localhost:8000/user/#/passwordreset/8586035765/374992445640`
1. Follow that link
1. Happy-path: Reset your password using the dialogue, then try logging-in again, but using the hash username `9ff4e87a0673406191ffe8b0a37777`, not your phone number! (phone isn't being swapped out for real username behind the scenes)
1. Non-happy path: a) Go back to the link after resetting password, b) change some of the parameters (e.g. phone number and token) in the URL. The UI will give you a generic "Token Expired" message (same message even if there was no token to begin with).

Other details:
1. Validations for text fields
1. Throttle network tab and make sure forms disable when needed

DXCanas commented:
I did a rebase on top of the ported old functionality - I don't know why it attaches my face to your commits, but could be worth looking at? https://github.com/DXCanas/kolibri-instant-schools-plugin-1/tree/reset-password-jb

jonboiser commented:
@DXCanas I think that's a newer GitHub feature. I'm not sure what it means, but maybe it means that I authored the commit, but you brought it into the code through the PR.

DXCanas commented:
ah got it. If you prefer, I can just strip out any modal logic. 

There used to be a modal that did nothing but display an iframe, which is why a modal was included in the port. I saw that we're moving away from the iframe approach, so I just put in some dummy templating.

jonboiser commented:
@DXCanas Yeah, maybe just remove the modal stuff from the port. It will probably eliminate/mitigate any eventual merge conflicts.

christianmemije commented on kolibri_instant_schools_plugin/assets/src/views/reset-password-modal/index.vue:
```diff
@@ -0,0 +1,120 @@
+    @cancel="closeModal"
+  >
+    <div class="contents">
+      <status
```
Instead of having a modal for `Account not found` it might be better if we just use a `ui-alert`. Also it does not help that when you click back, the field is empty, so you cannot see the inputted phone number.

christianmemije commented on kolibri_instant_schools_plugin/assets/src/views/reset-password-modal/index.vue:
```diff
@@ -0,0 +1,120 @@
+            const { code } = err.status;
+            if (code === 400) {
+              this.status = STATES.ACCOUNT_NOT_FOUND;
+            } else if (code === 500) {
```
Could there ever be a chance of a different error code?

jonboiser replied:
Not based on the implementation, unless Django will randomly send a 5xx error. I could make this a catch all instead of checking specifically for 500.

christianmemije commented on kolibri_instant_schools_plugin/assets/src/views/reset-password-page/new-password-form.vue:
```diff
@@ -0,0 +1,129 @@
+      },
+      newPwConfirmShouldValidate() {
+        return (
+          (this.newPw !== '' && this.newPwBlurred && this.newPwConfirm !== '') || this.formSubmitted
```
Missing `newPwConfirmBlurred`?

jonboiser replied:
Hm, I kind of wanted to avoid invalidating the `newPwConfirm` just for being blank (and thus requiring a different invalidText for that scenario). Here, it only invalidates if it doesn't match the `newPw`. This is covers when `newPwConfirm === ''` and `newPw !== ''`

christianmemije commented on kolibri_instant_schools_plugin/assets/src/views/reset-password-page/status.vue:
```diff
@@ -0,0 +1,94 @@
+      },
+    },
+    $trs: {
+      confirmingToken: 'Confirming the validity of our password reset token…',
```
'the' instead of 'our'? Also, would users even know what a token is?

jonboiser replied:
Haha, another intermediate state that wasn't specced out. Maybe instead of "token", it could be "link", since there is a "link expired" page.

christianmemije commented on kolibri_instant_schools_plugin/assets/src/views/sign-in-page/index.vue:
```diff
@@ -437,4 +453,8 @@
 
+  .reset-pw
+    margin-top: 1.5em
+    margin-bottom: -1em
```
Let's instead adjust the top margin of the divider.

jonboiser commented:
@christianmemije 

1. Adjusted divider margin-top
1. Phone-not-found error is shown inline in phone number form
1. Changed copy about token to reference "link"

christianmemije commented on kolibri_instant_schools_plugin/assets/src/views/reset-password-page/status.vue:
```diff
@@ -51,12 +51,12 @@
           case STATES.LINK_EXPIRED:
             return this.$tr('needToRequestNewToken');
-          case OTHER_ERROR:
+          case STATES.OTHER_ERROR:
```
Nice catch

jonboiser replied:
flying blind eslint-less in this project!
